### PR TITLE
Feature: Embed Donation Form button

### PR DESCRIPTION
**Pull request by alaca**
## Description

This PR adds a new feature to the Form Builder - Embed Button. This button allows folks to embed the form on a new or existing page. It also provides a pre-generated shortcode that the user can copy and embed by himself. 

## Affects
Form Builder

## Visuals

Screen Capture_select-area_20231106121244

## Testing Instructions

1. Add a few new pages/posts
2. Go to the Edit form page and click on the new Embed button
3. Select page/post from the dropdown and insert Form into it
4. Copy the shortcode and paste it to some post/page
5. Verify that everything is working on the frontend and that form is inserted into posts/pages

## Pre-review Checklist

-   [x] Relevant `@unreleased` tags included in DocBlocks
-   [x] [Self Review](https://give.gitbook.io/development-manual/devops/github/code-reviews#self-review) of code and UX completed

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1204808773418924

### CODE REVIEW

File: src/FormBuilder/resources/js/form-builder/src/components/EmbedForm/index.tsx [hunk 1]
```diff
@@ -0,0 +1,542 @@
...
+        });
+
+        if (data) {
+            data?.forEach(page => {
+                filtered.push({
+                    value: page.id,
+                    label: page.title.raw || __('(no title)', 'give'),
+                    content: page.content.raw,
+                    disabled: page.content.raw.includes(`wp:give/donation-form {"id":${formId}`), // disable pages that already have this form block included
+                });
```
Comment (jonwaldstein) [inline src/FormBuilder/resources/js/form-builder/src/components/EmbedForm/index.tsx:123]: can you explain why we need to "disable pages that already have this form block included"?

File: src/DonationForms/V2/ServiceProvider.php [hunk 2]
```diff
@@ -47,14 +47,7 @@ public function boot()
...
         // Onboarding
         Hooks::addAction('submitpost_box', DonationFormsAdminPage::class, 'renderMigrationGuideBox');
-
-        if (DonationFormsAdminPage::isShowingEditV2FormPage()) {
-            Hooks::addAction('admin_enqueue_scripts', DonationFormsAdminPage::class, 'loadEditFormScripts');
-        }
-
-        if (DonationFormsAdminPage::isShowingAddV2FormPage()) {
-            Hooks::addAction('admin_enqueue_scripts', DonationFormsAdminPage::class, 'loadAddFormScripts');
-        }
```
Comment (jonwaldstein) [inline src/DonationForms/V2/ServiceProvider.php:57]: Why are these being removed?

File: src/FormBuilder/resources/js/form-builder/src/components/EmbedForm/index.tsx [hunk 3]
```diff
@@ -0,0 +1,542 @@
...
+        }
+
+        setState(prevState => {
+            return {
+                ...prevState,
+                isCreating: true,
+            };
+        });
+
+        const response = await saveEntityRecord('postType', state.createPostType, {
```
Comment (jonwaldstein) [inline src/FormBuilder/resources/js/form-builder/src/components/EmbedForm/index.tsx:287]: Wow, it's pretty cool how much we can accomplish with just the the js wp api. 🪄

**Review by jonwaldstein — COMMENTED** [review]
@alaca this is turning out awesome! just had a few questions but overall i'm impressed 🤩 

Also, I don't know if this related but the form block (I thought) was supposed to have other options instead of just changing the form.  Honestly haven't looked at it in a while so maybe i'm wrong.  I also found it hard to select the block within a gutenberg - again, likely unrelated.

Screenshot 2023-11-14 at 1 21 49 PM

File: src/FormBuilder/resources/js/form-builder/src/components/EmbedForm/index.tsx (hunk 1 above)
Comment (alaca) [reply src/FormBuilder/resources/js/form-builder/src/components/EmbedForm/index.tsx:123]: I thought that having the two identical forms was something that nobody needs, but I may be wrong. We can easily remove that.

File: src/DonationForms/V2/ServiceProvider.php (hunk 2 above)
Comment (alaca) [reply src/DonationForms/V2/ServiceProvider.php:57]: This is related to migration onboarding stuff. The problem was that the logic to check if the current page is a donation form page was not functioning correctly and migration onboarding JS was loaded on regular pages causing JS errors in the console. The error was annoying so I had to fix it :smile:

File: src/FormBuilder/resources/js/form-builder/src/components/EmbedForm/index.tsx (hunk 3 above)
Comment (alaca) [reply src/FormBuilder/resources/js/form-builder/src/components/EmbedForm/index.tsx:287]: I know. It's amazing :smile:

**Comment by alaca** [comment]
@jonwaldstein you are right, there is much more option, but for legacy form. In @jdghinson's designs, there is nothing about that. Regarding block selection, Joshua's [PR](https://github.com/impress-org/givewp/pull/7082) will resolve that.

File: src/DonationForms/V2/ServiceProvider.php (hunk 2 above)
Comment (jonwaldstein) [reply src/DonationForms/V2/ServiceProvider.php:57]: Sounds good, thanks for the clarification.  Let's just make sure to add this to the QA testing instructions.

File: src/FormBuilder/resources/js/form-builder/src/components/EmbedForm/index.tsx (hunk 1 above)
Comment (jonwaldstein) [reply src/FormBuilder/resources/js/form-builder/src/components/EmbedForm/index.tsx:123]: Great thoughts! I appreciate the cleverness here, I would just be concerned people wouldn't understand this without an explanation so maybe better to remove for now.

Comment (alaca) [reply src/FormBuilder/resources/js/form-builder/src/components/EmbedForm/index.tsx]: Removed

**Comment by alaca** [comment]
@jonwaldstein, new changes have been added to match design requirements, can you review this again? Thanks!

File: src/DonationForms/Blocks/DonationFormBlock/Controllers/BlockRenderController.php [hunk 4]
```diff
@@ -38,12 +38,13 @@ public function render(array $attributes)
...
 
         $viewUrl = $this->getViewUrl($donationForm, $embedId);
+        $formUrl = add_query_arg(['p' => $blockAttributes->formId], site_url('?post_type=give_forms'));
 
         /**
          * Note: iframe-resizer uses querySelectorAll so using a data attribute makes the most sense to target.
          * It will also generate a dynamic ID - so when we have multiple embeds on a page there will be no conflict.
          */
-        return "<div class='root-data-givewp-embed' data-src='$viewUrl' data-givewp-embed-id='$embedId' data-form-format='$blockAttributes->formFormat' data-open-form-button='$blockAttributes->openFormButton'></div>";
+        return "<div class='root-data-givewp-embed' data-form-url='$formUrl' data-src='$viewUrl' data-givewp-embed-id='$embedId' data-form-format='$blockAttributes->formFormat' data-open-form-button='$blockAttributes->openFormButton'></div>";
```
Comment (jonwaldstein) [inline src/DonationForms/Blocks/DonationFormBlock/Controllers/BlockRenderController.php:47]: Can we actually use `esc_url()` for the data-attribute?  Technically `site_url` can be filtered so better to be safe here.

**Review by jonwaldstein — CHANGES_REQUESTED** [review]

File: src/DonationForms/Blocks/DonationFormBlock/Controllers/BlockRenderController.php (hunk 4 above)
Comment (alaca) [reply src/DonationForms/Blocks/DonationFormBlock/Controllers/BlockRenderController.php:47]: Resolved https://github.com/impress-org/givewp/pull/7080/commits/f97f48ca37a997188cb6989f06d8dddc4dc2f79a

**Review by jonwaldstein — APPROVED** [review]
Great job!  ready for QA 🌈

**Comment by alaca** [comment]
Plugin ZIP for QA https://github.com/impress-org/givewp/suites/18339702175/artifacts/1061269193

**Review by rickalday — APPROVED** [review]
Passed manual QA tests.

**Pull request merged**